lib/fputs.js

```diff
@@ -0,0 +1,185 @@
+        // it is assumed that all writes end with a newline (not checked)
+        var nchunks = this.datachunks.length;
+        if (nchunks > 0 && this.datachunks[nchunks-1].length + str.length <= this.writesize) {
+            // the last chunk has space for more
+            this.datachunks[nchunks-1] += str;
+        }
+        else {
+            // else start a new chunk
+            this.datachunks.push(str);
+        }
+
+        if (this.writtenLength && this.writtenLength >= this.unwrittenLength) {
+            // reset the lengths to avoid numeric overflow
+            this.writtenLength = this.unwrittenLength = 0;
+            this.resetCount += 1;
+        }
+        this.unwrittenLength += str.length;
+
+        if (!this._syncing) {
+            // if not currently syncing, start the sync thread
+            var self = this;
+            setTimeout(function(){ self._sync(); }, 1);
+            this._syncing = true;
+        }
+
+        if (callback) callback(null, str.length);
+        return this.unwrittenLength - this.writtenLength <= this.highWaterMark;
+    }
+
+    /**
+     * Wait until willing to accept more data.
+     * Waiting is optional, writes are always buffered.
+     */
+    Fputs.prototype.drain = function drain( maxUnwritten, callback ) {
+        if (this._error) return callback(this.returnError());
+
+        if (!callback) {
+            // The built-in heuristic is to buffer up to twice as much data
+            // as is written in a single write.
+            callback = maxUnwritten;
+            maxUnwritten = 2 * this.writesize;
+        }
+
+        if (this.unwrittenLength - this.writtenLength <= maxUnwritten) {
+            callback();
+        }
+        else {
+            var self = this;
+            setTimeout(function(){ self.drain(maxUnwritten, callback); }, 5);
+        }
+    }
+
+    /**
+     * Wait until all the data that has been written so far has been sent.
+     */
+    Fputs.prototype.fflush = function fflush( callback ) {
+        if (this._error) return callback(this.returnError());
+        var expectedWrittenLength = this.unwrittenLength;
+        var self = this;
+        var resetCount = this.resetCount;
+
+        (function waitloop() {
+            if (self.writtenLength >= expectedWrittenLength || self.resetCount != resetCount) {
+                callback(self.returnError());
+            }
+            else setTimeout(waitloop, 1);
+        })();
+    }
+
+    // the sync thread runs whenever there is data waiting,
+    // and tries to write chunks ending on line boundaries
+    Fputs.prototype._sync = function _sync( ) {
+        if (this.datachunks.length <= 0) {
+            this._syncing = false;
+            return;
+        }
+
+        var chunk = this.datachunks.shift();
+        var self = this;
+        this.writable.write(chunk, function(err, ret) {
+            self.writtenLength += chunk.length;
+            if (err) self.reportError(err);
```

Comment:
correct.  `reportError` remembers the error to deliver to the next `fflush` or `drain`.
Because writing the buffered data happens completely detached from the caller,
there is no-one to report errors to, so qfputs tries to carry on as best it can.